### PR TITLE
feat(hilla): make tsconfig.json React compatible

### PR DESCRIPTION
**Pull request by platosha**
Summary of changes:

- Added a good default for `jsx` setting
- Relaxed file extensions for TypeScript, so that `.tsx` and such are considered

This should not interfere with existing TypeScript / Lit-based code.

### CODE REVIEW

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 1]
```diff
@@ -5,11 +5,13 @@
...
 {
   "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
     "sourceMap": true,
+    "jsx": "react-jsx",
     "inlineSources": true,
     "module": "esNext",
     "target": "es2020",
-    "moduleResolution": "node",
+    "moduleResolution": "node16",
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:14]: What is the difference between "node" and "node16"? We already require node 16

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 2]
```diff
@@ -5,11 +5,13 @@
 // For more information about the configurations, please refer to http://www.typescriptlang.org/docs/handbook/tsconfig-json.html
 {
   "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
     "sourceMap": true,
+    "jsx": "react-jsx",
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:10]: Does this have any impact when you are not using jsx?

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 3]
```diff
@@ -5,11 +5,13 @@
 // For more information about the configurations, please refer to http://www.typescriptlang.org/docs/handbook/tsconfig-json.html
 {
   "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:8]: Is this needed for React or just handy?

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 4]
```diff
@@ -19,6 +21,7 @@
     "noUnusedLocals": false,
     "noUnusedParameters": false,
     "experimentalDecorators": true,
+    "importsNotUsedAsValues": "error",
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:24]: How is this related to React?

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 5]
```diff
@@ -27,8 +30,7 @@
     }
   },
   "include": [
-    "frontend/**/*.ts",
-    "frontend/index.js",
+    "frontend/**/*",
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:33]: This will now include images, json, everything. Is that the intent or are we mainly looking to add `*.tsx` here?

**Review by Artur- — COMMENTED** [review]
Please clarify the changes so we know in the future why they were made

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 1 above)
Comment (Artur-) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:14]: Error:  frontend/generated/index.ts(20,25): error TS2835: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean '../../target/frontend/generated-flow-imports.js'?

At least it is not 100% compatible

Comment (platosha) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:14]: https://github.com/microsoft/TypeScript/issues/35290

enables a very handy `exports` definition in `package.json` in dependencies, which is already supported by Vite but confusingly not by TypeScript... We use it in `@hilla/react-components`

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 5 above)
Comment (platosha) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:33]: yes

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 4 above)
Comment (platosha) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:24]:
```suggestion
    "experimentalDecorators": true,
```

Comment (platosha) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:24]: not related it seems,

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 6]
```diff
@@ -5,11 +5,13 @@
 // For more information about the configurations, please refer to http://www.typescriptlang.org/docs/handbook/tsconfig-json.html
 {
   "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
     "sourceMap": true,
```
Comment (platosha) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:9]:
```suggestion
    "sourceMap": true,
```

seems not needed, as `import * as React from 'react';` is preferred

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 1 above)
Comment (Lodin) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:14]: It also defaults to the Node 16+ ESM/CJS resolution algorithm which allows importing CJS from ESM without any specific interop.

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 2 above)
Comment (platosha) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:10]: It’s a nice tweak, the default is "react" option, which requires a full `React` import. See https://www.typescriptlang.org/tsconfig#jsx

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 1 above)
Comment (Lodin) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:13]: > At least it is not 100% compatible

Unfortunately, yes. Still, I believe this is very important change; it looks like to become the standard de facto, so we need to push it

Comment (Artur-) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:13]: Yeah but how much does it break? Can we do it for 23.3 or only for 24?

Comment (Artur-) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:13]: I get 38 errors from the default Hilla project when enabling this

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 5 above)
Comment (Lodin) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:31]: I would suggest to keep it this way. Otherwise, if the user wants to include image/css/other file to the bundle (as `base64` string or as an automatic link), the Typescript will show an error even if we have a module declaration in the `types.d.ts` file (as far as I remember).

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json (hunk 1 above)
Comment (Lodin) [reply flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:13]: Yeah, probably, it's better to wait for v24.

**Review by Artur- — APPROVED** [review]

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 7]
```diff
@@ -7,6 +7,7 @@
   "flow_version": "23.3.0",
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:7]: This needs to be updated

**Comment by mshabarov** [comment]
@Artur- @platosha yes, we should update some parts to make autoupdate works fine:
1. Increment version in tsconfig.json ("flow_version" property): can be `"flow_version" : "23.3.1"`.
2. Create a new tsconfig-v23.3.0.json template in flow-server resources and put the old content there (before your patch)
3. Update `vaadinVersion` array in `TaskGenerateTsConfig` with `v23.3.0`
4. Put a new content in `tsconfig-reference.json` in tests.

P.S. yes, I know this manual work looks ugly...maybe I can remove (2) - (4) someday.

**Comment by platosha** [comment]
@Artur- @mshabarov OK done, even though `23.3.0` was not even released yet...

**Comment by platosha** [comment]
Whoa, a test failed.

@mshabarov FYI the instructions were missing a step: https://github.com/vaadin/flow/pull/15084/commits/9af2f3cb6229178c422a1197103cfbe7e5488930

File: flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json [hunk 8]
```diff
@@ -4,9 +4,10 @@
 // You might want to change the configurations to fit your preferences
 // For more information about the configurations, please refer to http://www.typescriptlang.org/docs/handbook/tsconfig-json.html
 {
-  "flow_version": "23.3.0",
+  "flow_version": "23.3.1",
```
Comment (Artur-) [inline flow-server/src/main/resources/com/vaadin/flow/server/frontend/tsconfig.json:7]: Won't this cause confusion forever as it is not part of 23.3.1 but 23.30? Can it be 23.3.0.1 instead?

**Review by Artur- — APPROVED** [review]

**Review by mshabarov — APPROVED** [review]
